expression/integration_test.go

```diff
@@ -8938,3 +8939,126 @@ func (s *testIntegrationSerialSuite) TestCollationForBinaryLiteral(c *C) {
+		"(6, null);")
+	tk.MustExec("insert into t_blob values " +
+		"(1, x'00000007128243F7'), " +
+		"(2, x'07128243F7'), " +
+		"(3, null);")
+	tk.MustExec("insert into t_varchar values " +
+		"(1, 'abcd'), " +
+		"(2, null);")
+
+	// Integers
+	tk.MustQuery("select hex(vitess_hash(a)) from t_int").
+		Check(testkit.Rows(
+			"31265661E5F1133",
+			"31B565D41BDF8CA",
+			"1EFD6439F2050FFD",
+			"355550B2150E2451",
+			"1E1788FF0FDE093C",
+			"<nil>"))
+
+	// Nested function sanity test
+	tk.MustQuery("select hex(vitess_hash(convert(a, decimal(8,4)))) from t_int where id = 5").
+		Check(testkit.Rows("1E1788FF0FDE093C"))
+
+	// Blobs
+	tk.MustQuery("select hex(vitess_hash(a)) from t_blob").
+		Check(testkit.Rows(
+			"31265661E5F1133", // Matches vitess_hash(30375298039) from above
+			"31265661E5F1133",
+			"<nil>"))
+
+	tk.MustQuery("select hex(vitess_hash(a)) from t_varchar").
+		Check(testkit.Rows(
+			"7CB9A2D545189781",
+			"<nil>"))
+}
+
+func (s *testIntegrationSuite) TestVitessHashMatchesVitessShards(c *C) {
+	defer s.cleanEnv(c)
+	tk := testkit.NewTestKit(c, s.store)
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.Experimental.AllowsExpressionIndex = true
+	})
+
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t;")
+	tk.MustExec("create table t(customer_id bigint, id bigint, expected_shard bigint unsigned, computed_shard bigint unsigned null, primary key (customer_id, id));")
+	tk.MustExec("create index t_vitess_shard ON t((vitess_hash(customer_id) >> 56));")
+
+	tk.MustExec("insert into t (customer_id, id, expected_shard) values " +
+		"(30370720100, 1, x'd6'), " +
+		"(30370670010, 2, x'd6'), " +
+		"(30370689320, 3, x'e1'), " +
+		"(30370693008, 4, x'e0'), " +
+		"(30370656005, 5, x'89'), " +
+		"(30370702638, 6, x'89'), " +
+		"(30370658809, 7, x'ce'), " +
+		"(30370665369, 8, x'cf'), " +
+		"(30370706138, 9, x'85'), " +
+		"(30370708769, 10, x'85'), " +
+		"(30370711915, 11, x'a3'), " +
+		"(30370712595, 12, x'a3'), " +
+		"(30370656340, 13, x'7d'), " +
+		"(30370660143, 14, x'7c'), " +
+		"(30371738450, 15, x'fc'), " +
+		"(30371683979, 16, x'fd'), " +
+		"(30370664597, 17, x'92'), " +
+		"(30370667361, 18, x'93'), " +
+		"(30370656406, 19, x'd2'), " +
+		"(30370716959, 20, x'd3'), " +
+		"(30375207698, 21, x'9a'), " +
+		"(30375168766, 22, x'9a'), " +
+		"(30370711813, 23, x'ca'), " +
+		"(30370721803, 24, x'ca'), " +
+		"(30370717957, 25, x'97'), " +
+		"(30370734969, 26, x'96'), " +
+		"(30375203572, 27, x'98'), " +
+		"(30375292643, 28, x'99'); ")
+
+	// Sanity check the shards being computed correctly
+	tk.MustExec("update t set computed_shard =  (vitess_hash(customer_id) >> 56);")
+
+	tk.MustQuery("select customer_id, id, hex(expected_shard), hex(computed_shard) from t where expected_shard <> computed_shard").
+		Check(testkit.Rows())
+
+	// Test the correct usage of index
+	tk.MustQuery(`explain format = 'brief' select customer_id from t where (vitess_hash(customer_id) >> 56) = x'd6' ORDER BY id`).
```

Comment:
1. it's better to add these tests to `explaintest`.
2. could you add a test to ensure whether the `vitess_hash()` function is foldable in the constant folding phase? a test example: `select vitess_hash(123) from t`, the execution plan should show a constant rather than a `vitess_hash(123)` function.

Reply:
Done and done.